packages/protobuf/src/next/to-json.ts

```diff
@@ -0,0 +1,486 @@
+  opts: JsonWriteOptions,
+): JsonValue | undefined {
+  if (!msg.desc.typeName.startsWith("google.protobuf.")) {
+    return undefined;
+  }
+  switch (msg.desc.typeName) {
+    case "google.protobuf.Any":
+      return anyToJson(msg.message as Any, opts);
+    case "google.protobuf.Timestamp":
+      return timestampToJson(msg.message as Timestamp);
+    case "google.protobuf.Duration":
+      return durationToJson(msg.message as Duration);
+    case "google.protobuf.FieldMask":
+      return fieldMaskToJson(msg.message as FieldMask);
+    case "google.protobuf.Struct":
+      return structToJson(msg.message as Struct, opts);
+    case "google.protobuf.Value":
+      return valueToJson(msg.message as Value, opts);
+    case "google.protobuf.ListValue":
+      return listValueToJson(msg.message as ListValue, opts);
+    default:
+      if (isWktWrapperDesc(msg.desc)) {
+        const valueField = msg.desc.fields[0];
+        return scalarToJson(valueField.scalar, msg.get(valueField));
+      }
+      return undefined;
+  }
+}
+
+function anyToJson(val: Any, opts: JsonWriteOptions): JsonValue {
+  if (val.typeUrl === "") {
+    return {};
+  }
+  const message = opts.descSet ? anyUnpack(val, opts.descSet) : undefined;
+  if (!message) {
+    throw new Error(
+      `cannot encode message ${val.$typeName} to JSON: "${val.typeUrl}" is not in the type registry`,
+    );
+  }
+  let json = reflectToJson(reflect(message), opts);
+  if (
+    message.$desc.typeName.startsWith("google.protobuf.") ||
+    json === null ||
+    Array.isArray(json) ||
+    typeof json !== "object"
+  ) {
+    json = { value: json };
+  }
+  json["@type"] = val.typeUrl;
+  return json;
+}
+
+function durationToJson(val: Duration) {
+  if (
+    Number(val.seconds) > 315576000000 ||
+    Number(val.seconds) < -315576000000
+  ) {
+    throw new Error(
+      `cannot encode ${val.$typeName} to JSON: value out of range`,
+    );
+  }
+  let text = val.seconds.toString();
+  if (val.nanos !== 0) {
+    let nanosStr = Math.abs(val.nanos).toString();
+    nanosStr = "0".repeat(9 - nanosStr.length) + nanosStr;
+    if (nanosStr.substring(3) === "000000") {
+      nanosStr = nanosStr.substring(0, 3);
+    } else if (nanosStr.substring(6) === "000") {
+      nanosStr = nanosStr.substring(0, 6);
+    }
+    text += "." + nanosStr;
+    if (val.nanos < 0 && Number(val.seconds) == 0) {
+      text = "-" + text;
+    }
+  }
+  return text + "s";
+}
+
+function fieldMaskToJson(val: FieldMask) {
+  return val.paths
+    .map((p) => {
+      // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+      if (p.match(/_[0-9]?_/g) || p.match(/[A-Z]/g)) {
+        throw new Error(
+          `cannot encode ${val.$typeName} to JSON: lowerCamelCase of path name "` +
+            p +
+            '" is irreversible',
+        );
+      }
+      return protoCamelCase(p);
+    })
+    .join(",");
+}
+
+function structToJson(val: Struct, opts: JsonWriteOptions) {
```

Comment:
It looks like we can remove `opts` from here, and also from `valueToJson` and from `listValueToJson`.